examples/nidcpower_source_dc_voltage/measurement.py

```diff
@@ -78,89 +83,97 @@ def measure(
-            time_remaining = measurement_service.context.time_remaining
-
-            channels.source_mode = nidcpower.SourceMode.SINGLE_POINT
-            channels.output_function = nidcpower.OutputFunction.DC_VOLTAGE
-            channels.current_limit = current_limit
-            channels.voltage_level_range = voltage_level_range
-            channels.current_limit_range = current_limit_range
-            channels.source_delay = hightime.timedelta(seconds=source_delay)
-            channels.voltage_level = voltage_level
-            # The Measurement named tuple doesn't support type annotations:
-            # https://github.com/ni/nimi-python/issues/1885
-            measured_values = []
-            with channels.initiate():
-                deadline = time.time() + time_remaining
-                while True:
-                    if time.time() > deadline:
-                        measurement_service.context.abort(
-                            grpc.StatusCode.DEADLINE_EXCEEDED, "deadline exceeded"
-                        )
-                    if pending_cancellation:
-                        measurement_service.context.abort(
-                            grpc.StatusCode.CANCELLED, "client requested cancellation"
-                        )
-                    try:
-                        channels.wait_for_event(nidcpower.enums.Event.SOURCE_COMPLETE, timeout=0.1)
-                        break
-                    except nidcpower.errors.DriverError as e:
-                        """
-                        There is no native way to support cancellation when taking a DCPower
-                        measurement. To support cancellation, we will be calling WaitForEvent
-                        until it succeeds or we have gone past the specified timeout. WaitForEvent
-                        will throw an exception if it times out, which is why we are catching
-                        and doing nothing.
-                        """
-                        if (
-                            e.code == NIDCPOWER_WAIT_FOR_EVENT_TIMEOUT_ERROR_CODE
-                            or e.code == NIDCPOWER_TIMEOUT_EXCEEDED_ERROR_CODE
-                        ):
-                            pass
-                        else:
-                            raise
-
-                measured_values = channels.measure_multiple()
-                for index, mapping in enumerate(channel_mappings):
-                    measured_values[index] = measured_values[index]._replace(
-                        in_compliance=session.channels[mapping.channel].query_in_compliance()
-                    )
-            session = None  # Don't abort after this point
-
-    _log_measured_values(channel_mappings, measured_values)
+        for pin_index, pin_name in enumerate(pin_names):
+            for session_info, channel_mapping in reservation.get_pin_based_session_info(pin_name):
+                with create_session(
```

Comment:
Before, this example sourced and measured multiple pins & sites in parallel.

Now, it sources and measures one pin+site at a time. This makes the measurement slower for no reason. 

There are two other examples that demonstrate how to create multiple sessions. Note that they loop over session as the **innermost** loop, not the **outermost** loop:
- [nifgen_standard_function](https://github.com/ni/measurementlink-python/blob/main/examples/nifgen_standard_function/measurement.py) 
- [niswitch_control_relays](https://github.com/ni/measurementlink-python/blob/main/examples/niswitch_control_relays/measurement.py)

Multi-session examples should manage a list of sessions, with an ExitStack for cleanup. Initiate could use a separate ExitStack if that makes more sense.

Reply:
Here's what takes significant time:
- create_session:
  - In InstrumentStudio, this creates a new NI-DCPower channel expansion session. This has significant overhead, especially for simulated devices. I think it has time complexity of O(Nchannels), so repeatedly creating/destroying the same session for each channel is O(Nchannels^2) with a relatively large multiplier. You should call create_session() once per channel group, and prefer fewer channel groups except when necessary (e.g. for multi-site parallelism).
  - In TestStand, this attaches to the existing session that was created in setup, which is O(1) and fast.
- initiate:
  - This starts the source operation, which is pretty fast.
  - **IMPORTANT**: If you initiate multiple sessions or channels in a row, the SMU hardware will perform source operations **simultaneously**.
  - **ALSO IMPORTANT**: If you initiate multiple channels in a single call to a single session, I think NI-DCPower automatically dispatches to multiple threads. This applies to most of these API calls, not just initialize.
- wait_for_event:
  - This waits for the source operation to complete. Part of this time is hardware overhead, part is the `source_delay` setting. If you set `source_delay` to 5 seconds, the time from the beginning of initiate() to the end of wait_for_event() is a little longer than 5 seconds.
    - Note: simulated devices don't simulate this delay accurately.
  - If you set `source_delay` to 5 seconds, initiate() N times, and wait_for_event() N times, it will take a little longer than 5 seconds.
  - If you set `source_delay` to 5 seconds and call initiate() + wait_for_event() in a single loop, it will take N * 5 seconds.
- measure:
  - This performs a measurement, which has some hardware overhead.
- abort:
  - You don't see this call in the code, because it's handled by the context manager returned from initialize().
  - I think this is pretty fast.
- close session:
  - You don't see this call in the code because it's handled by the context manager returned from create_session.
  - In InstrumentStudio, this probably has significant overhead, like create_session.